coresdk/src/backend/concurrency_utils.h

```diff
@@ -0,0 +1,103 @@
+//
+//  concurrency_driver.h
+//  splashkit
+//
+//  Created by James Armstrong on 21/08/2016
+//  Copyright © 2016 Andrew Cain. All rights reserved.
+//
+
+#ifndef sgsdl2_SGSDL2ConcurrencyUtils_h
+#define sgsdl2_SGSDL2ConcurrencyUtils_h
+
+#include <mutex>
+#include <thread>
+#include <condition_variable>
+#include <queue>
+
+using namespace std;
+
+class semaphore
+{
+private:
+    mutex _mutex;
+    condition_variable _cv;
+    int _tokens;
+
+public:
+    semaphore()
+    {
+        _tokens = 0;
+    }
+
+    semaphore(int tokens)
+    {
+        _tokens = tokens;
+    }
+
+    void acquire()
+    {
+        unique_lock<mutex> lock(_mutex);
+        while (_tokens <= 0)
+        {
+            _cv.wait(lock);
+        }
+        _tokens--;
+    }
+
+    void release(int num = 1)
+    {
+        lock_guard<mutex> lock(_mutex);
+        _tokens += num;
+        _cv.notify_all();
+    }
+};
+
+template <typename T>
+class channel
+{
+private:
+    queue<T> _queue;
+    semaphore _take_permission;
+    mutex _mutex;
+
+    T dequeue()
+    {
+        T data = _queue.front();
+        _queue.pop();
+        return data;
+    }
+
+public:
+    void put(T data)
+    {
+        unique_lock<mutex> lock(_mutex);
+        _queue.push(data);
+        lock.unlock();
+
+        _take_permission.release();
+    }
+
+    T take()
+    {
+        _take_permission.acquire();
+
+        lock_guard<mutex> lock(_mutex);
+        return dequeue();
+    }
+
+    bool try_take(T& data)
+    {
+        lock_guard<mutex> lock(_mutex);
```

Comment:
I think there is a race condition here...

I suggest adding a try_acquire to the semaphore and trying to get permission from the _take_permission here instead.